jsonparser/tidwall.go

```diff
@@ -0,0 +1,373 @@
+		return 0, ErrNotOfExpectedType
+	}
+
+	return result.Float(), nil
+}
+
+// GetFloatOrZero retrieves a float value for a given key from JSON bytes.
+// If the key does not exist, json is invalid or value is not parsable to a float, it returns 0.
+// e.g `GetFloatOrZero({"key": 123.456}, "key1") returns 0.
+//
+//	`GetFloatOrZero({"key": 123.456}, "key")` returns 123.456.
+//	`GetFloatOrZero({"key": "123.456"}, "key")` returns 123.456 as `"123.456"` is parsed to 123.456.
+//	`GetFloatOrZero({"key": "val"}, "key")` returns 0 as `val` is not parsable to float.
+//	`GetFloatOrZero({"key": true}, "key")` returns 1 as `true` is parsed to 1.
+func (p *tidwallJSONParser) GetFloatOrZero(data []byte, path ...string) float64 {
+	if len(data) == 0 || len(path) == 0 {
+		return 0
+	}
+
+	if path[0] == "" {
+		return 0
+	}
+
+	// Join keys with dots to create a path
+	dotSeperatedPath := getPath(path...)
+
+	// Use gjson to get the value
+	result := gjson.GetBytes(data, dotSeperatedPath)
+
+	return result.Float()
+}
+
+// GetString retrieves a string value for a given key from JSON bytes using gjson
+func (p *tidwallJSONParser) GetString(data []byte, path ...string) (string, error) {
+	if len(data) == 0 {
+		return "", ErrEmptyJSON
+	}
+
+	if len(path) == 0 {
+		return "", ErrNoKeysProvided
+	}
+
+	key := path[0]
+	if key == "" {
+		return "", ErrEmptyKey
+	}
+
+	// Join keys with dots to create a path
+	dotSeperatedPath := getPath(path...)
+
+	// Use gjson to get the value
+	result := gjson.GetBytes(data, dotSeperatedPath)
+	if !result.Exists() {
+		return "", ErrKeyNotFound
+	}
+
+	// Check if the value is a string
+	if result.Type != gjson.String {
+		return "", ErrNotOfExpectedType
+	}
+
+	return result.String(), nil
+}
+
+// GetStringOrEmpty retrieves a string value for a given key from JSON bytes.
+// If the key does not exist, json is invalid or value is not parsable to a string, it returns an empty string.
+// e.g `GetStringOrEmpty({"key": "val"}, "key1")` returns "".
+//
+//	`GetStringOrEmpty({"key": "val"}, "key")` returns "val".
+//	`GetStringOrEmpty({"key": 123}, "key")` returns "123" as 123 is parsed to string.
+//	`GetStringOrEmpty({"key": true}, "key")` returns "true" as `true` is parsed to string.
+func (p *tidwallJSONParser) GetStringOrEmpty(data []byte, path ...string) string {
+	if len(data) == 0 || len(path) == 0 {
+		return ""
+	}
+
+	if path[0] == "" {
+		return ""
+	}
+
+	// Join keys with dots to create a path
+	dotSeperatedPath := getPath(path...)
+
+	// Use gjson to get the value
+	result := gjson.GetBytes(data, dotSeperatedPath)
+
+	return result.String()
+}
+
+// SetValue sets the value for a given key in JSON bytes using sjson
+func (p *tidwallJSONParser) SetValue(data []byte, value interface{}, path ...string) ([]byte, error) {
+	if len(data) == 0 {
+		// If data is empty, create a new JSON object
+		data = []byte("{}")
+	}
+
+	if len(path) == 0 {
+		return nil, ErrNoKeysProvided
+	}
+
+	if lo.Contains(path, "") {
```

Comment:
A small optimization that can be done here in `tidwall` is to check for empty keys while you're iterating to create the dot separated path. Instead you iterate through all keys first to check if one is empty and then you do another loop to iterate across the same data structure again to get `dotSeperatedPath`.